File: sdks/typescript/src/experiments.ts
import { createHash } from "node:crypto";

import { flush, getClient, HttpExporter } from "./client.js";
import { trace } from "./tracing.js";
import { randomHex } from "./util.js";

const DATASET_PAGE_SIZE = 5000;

export interface DatasetItem {
  input: unknown;
  expected?: unknown;
  metadata?: Record<string, unknown>;
  item_id?: string;
}

export interface EvaluationResult {
  name: string;
  score?: number | null;
  passed?: boolean | null;
  label?: string | null;
  comment?: string | null;
  metadata?: Record<string, unknown>;
}

export interface EvaluatorArgs {
  input: unknown;
  output: unknown;
  expected: unknown;
  metadata: Record<string, unknown>;
}

type EvaluatorReturn = EvaluationResult | number | boolean | Omit<EvaluationResult, "name"> | null | undefined;

export type EvaluatorFunction = ((args: EvaluatorArgs) => EvaluatorReturn | Promise<EvaluatorReturn>) & { evaluatorName?: string };
export interface EvaluatorObject {
  name: string;
  evaluate(args: EvaluatorArgs): EvaluatorReturn | Promise<EvaluatorReturn>;
}
export type Evaluator = EvaluatorFunction | EvaluatorObject;

export interface ItemResult {
  item_id: string;
  input: unknown;
  expected: unknown;
  output: unknown;
  error: string | null;
  status: "completed" | "failed";
  trace_id: string;
  duration_ms: number;
  scores: EvaluationResult[];
}

export interface ExperimentResult {
  experimentId: string;
  name: string;
  dataset: string | null;
  results: ItemResult[];
  summary: ExperimentSummary;
  persisted: boolean;
  url?: string;
  /** Mean score for one evaluator, or undefined if it produced no numeric scores. */
  score(name: string): number | undefined;
}

export interface ExperimentSummary {
  items: number;
  errors: number;
  error_rate: number;
  avg_latency_ms: number | null;
  scores: Record<string, { mean: number | null; min: number | null; max: number | null; count: number; pass_rate: number | null }>;
}

export interface RunExperimentOptions<I = any, O = any> {
  /** A dataset name or id on the AgentMesh server, or items inline. */
  dataset: string | DatasetItem[];
  /** Your agent or prompt under test. Receives each item's input. */
  task: (input: I, item: DatasetItem) => O | Promise<O>;
  evaluators?: Evaluator[];
  name?: string;
  description?: string;
  metadata?: Record<string, unknown>;
  maxConcurrency?: number;
}

/**
 * Run `task` over a dataset, score every output, and store the results as an experiment in AgentMesh.
 * Each item runs inside its own trace (tagged `experiment`).
 */
export async function runExperiment<I = any, O = any>(options: RunExperimentOptions<I, O>): Promise<ExperimentResult> {
  const client = getClient();
  const exporter = client.exporter instanceof HttpExporter ? client.exporter : undefined;
  let datasetName: string | null = null;
  let datasetId: string | null = null;
  let items: DatasetItem[];
  if (typeof options.dataset === "string") {
    if (!exporter) throw new Error("Loading a dataset by name needs the HTTP exporter (an AgentMesh endpoint)");
    const path = `/api/datasets/${encodeURIComponent(options.dataset)}`;
    items = [];
    // The server returns items a page at a time; keep reading until item_count is reached.
    for (;;) {
      const page = (await exporter.request("GET", `${path}?limit=${DATASET_PAGE_SIZE}&offset=${items.length}`)) as {
        dataset_id: string;
        name: string;
        item_count: number;
        items: DatasetItem[];
      };
      datasetName = page.name;
      datasetId = page.dataset_id;
      items.push(...(page.items ?? []));
      if (!page.items?.length || items.length >= Number(page.item_count ?? 0)) break;
    }
  } else {
    items = options.dataset;
  }
  items.forEach((item, index) => {
    if (!item || typeof item !== "object" || !("input" in item)) throw new Error(`dataset item ${index} needs an 'input'`);
  });

  const experimentId = `exp_${randomHex(8)}`;
  const name = options.name ?? `${datasetName ?? "experiment"}-${new Date().toISOString().replace(/[-:]/g, "").slice(0, 15)}`;
  const evaluators = options.evaluators ?? [];
  const header = {
    experiment_id: experimentId,
    name,
    description: options.description ?? null,
    dataset: datasetId,
    dataset_name: datasetName,
    status: "running",
    started_at: new Date().toISOString(),
    evaluators: evaluators.map(evaluatorName),
    metadata: { ...(options.metadata ?? {}), task: options.task.name || "task", sdk: "typescript" },
  };
  if (exporter) await exporter.request("POST", "/api/experiments", JSON.stringify({ ...header, results: [] }));

  const results: ItemResult[] = new Array(items.length);
  let next = 0;
  const worker = async () => {
    while (next < items.length) {
      const index = next;
      next += 1;
      results[index] = await runItem(items[index]!, options.task, evaluators, experimentId, name, datasetName);
    }
  };
  await Promise.all(Array.from({ length: Math.max(1, Math.min(options.maxConcurrency ?? 4, items.length || 1)) }, worker));
  await flush();

  for (const result of results) {
    for (const score of result.scores) {
      if ((score.score === null || score.score === undefined) && (score.passed === null || score.passed === undefined)) continue;
      client.score({
        trace_id: result.trace_id,
        name: score.name,
        value: score.score ?? score.passed ?? null,
        passed: score.passed ?? null,
        label: score.label ?? null,
        comment: score.comment ?? null,
        source: "experiment",
        metadata: { experiment_id: experimentId, item_id: result.item_id, ...(score.metadata ?? {}) },
      });
    }
  }
  await flush();

  let summary = summarize(results);
  if (exporter) {
    const final = { ...header, status: "completed", ended_at: new Date().toISOString() };
    for (let start = 0; start < Math.max(results.length, 1); start += 200) {
      const saved = (await exporter.request(
        "POST",
        "/api/experiments",
        JSON.stringify({ ...final, results: results.slice(start, start + 200) }),
      )) as { summary?: ExperimentSummary };
      if (saved?.summary) summary = saved.summary;
    }
  }
  return {
    experimentId,
    name,
    dataset: datasetName,
    results,
    summary,
    persisted: Boolean(exporter),
    url: exporter ? `${exporter.endpoint}/?page=datasets&experiment=${experimentId}` : undefined,
    score(evaluator: string) {
      return summary.scores[evaluator]?.mean ?? undefined;
    },
  };
}

async function runItem(
  item: DatasetItem,
  task: (input: any, item: DatasetItem) => unknown,
  evaluators: Evaluator[],
  experimentId: string,
  experimentName: string,
  datasetName: string | null,
): Promise<ItemResult> {
  const itemId = item.item_id ?? inlineItemId(item.input);
  const started = performance.now();
  const outcome = await trace(
    `experiment:${experimentName}`,
    { tags: ["experiment"], metadata: { experiment_id: experimentId, item_id: itemId, dataset: datasetName }, input: item.input },
    async (root) => {
      try {
        const output = await task(item.input, item);
        root.setOutput(output);
        return { traceId: root.traceId, output, error: null as string | null };
      } catch (error) {
        root.recordException(error);
        const message = error instanceof Error ? `${error.name}: ${error.message}` : String(error);
        return { traceId: root.traceId, output: undefined as unknown, error: message };
      }
    },
  );
  const result: ItemResult = {
    item_id: itemId,
    input: item.input,
    expected: item.expected ?? null,
    output: outcome.output ?? null,
    error: outcome.error,
    status: outcome.error ? "failed" : "completed",
    trace_id: outcome.traceId,
    duration_ms: Math.round((performance.now() - started) * 1000) / 1000,
    scores: [],
  };
  if (!outcome.error) {
    for (const evaluator of evaluators) {
      result.scores.push(await runEvaluator(evaluator, { input: item.input, output: outcome.output, expected: item.expected ?? null, metadata: item.metadata ?? {} }));
    }
  }
  return result;
}

export async function runEvaluator(evaluator: Evaluator, args: EvaluatorArgs): Promise<EvaluationResult> {
  const name = evaluatorName(evaluator);
  try {
    const value = typeof evaluator === "function" ? await evaluator(args) : await evaluator.evaluate(args);
    return coerceResult(name, value);
  } catch (error) {
    return { name, label: "error", comment: error instanceof Error ? `${error.name}: ${error.message}` : String(error) };
  }
}

function evaluatorName(evaluator: Evaluator): string {
  return typeof evaluator === "function" ? evaluator.evaluatorName ?? (evaluator.name || "evaluator") : evaluator.name;
}

function coerceResult(name: string, value: EvaluatorReturn): EvaluationResult {
  if (typeof value === "boolean") return { name, score: value ? 1 : 0, passed: value };
  if (typeof value === "number") return { name, score: value };
  if (value === null || value === undefined) return { name, label: "skipped" };
  const result = value as EvaluationResult;
  return { ...result, name: result.name ?? name };
}

function summarize(results: ItemResult[]): ExperimentSummary {
  const buckets = new Map<string, { values: number[]; passed: boolean[] }>();
  for (const result of results) {
    for (const score of result.scores) {
      const bucket = buckets.get(score.name) ?? { values: [], passed: [] };
      buckets.set(score.name, bucket);
      if (typeof score.score === "number") bucket.values.push(score.score);
      if (typeof score.passed === "boolean") bucket.passed.push(score.passed);
    }
  }
  const scores: ExperimentSummary["scores"] = {};
  for (const [name, { values, passed }] of buckets) {
    scores[name] = {
      mean: values.length ? values.reduce((a, b) => a + b, 0) / values.length : null,
      min: values.length ? Math.min(...values) : null,
      max: values.length ? Math.max(...values) : null,
      count: values.length,
      pass_rate: passed.length ? passed.filter(Boolean).length / passed.length : null,
    };
  }
  const errors = results.filter((result) => result.status === "failed").length;
  const durations = results.map((result) => result.duration_ms);
  return {
    items: results.length,
    errors,
    error_rate: results.length ? errors / results.length : 0,
    avg_latency_ms: durations.length ? durations.reduce((a, b) => a + b, 0) / durations.length : null,
    scores,
  };
}

/**
 * The id the Python SDK gives an inline item, so experiments from either language compare item by item.
 * Matches for strings, integers, booleans, null, arrays, and objects. Floats can differ (JavaScript cannot
 * tell `1.0` from `1`), so give items an explicit `item_id` when inputs contain them.
 */
export function inlineItemId(input: unknown): string {
  return `item_${createHash("sha1").update(pythonJson(input)).digest("hex").slice(0, 16)}`;
}

/** `json.dumps(value, sort_keys=True)` formatting: ", " and ": " separators, ASCII-only escapes. */
function pythonJson(value: unknown): string {
  if (value === null || value === undefined) return "null";
  if (typeof value === "boolean") return value ? "true" : "false";
  if (typeof value === "number") return Number.isInteger(value) ? String(value) : String(value);
  if (typeof value === "string") {
    // ensure_ascii: escape every UTF-16 code unit from DEL (0x7F) up as backslash-u plus four hex digits.
    const text = JSON.stringify(value);
    let escaped = "";
    for (let index = 0; index < text.length; index += 1) {
      const code = text.charCodeAt(index);
      escaped += code >= 0x7f ? `${String.fromCharCode(92)}u${code.toString(16).padStart(4, "0")}` : text[index];
    }
    return escaped;
  }
  if (Array.isArray(value)) return `[${value.map(pythonJson).join(", ")}]`;
  if (typeof value === "object") {
    const entries = Object.entries(value as Record<string, unknown>)
      .filter(([, item]) => item !== undefined)
      .sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0));
    return `{${entries.map(([key, item]) => `${pythonJson(key)}: ${pythonJson(item)}`).join(", ")}}`;
  }
  return JSON.stringify(String(value));
}

// -- evaluators ------------------------------------------------------------------------------

function normalize(value: unknown, caseSensitive: boolean): string {
  const text = typeof value === "string" ? value : JSON.stringify(value) ?? "";
  const collapsed = text.split(/\s+/).filter(Boolean).join(" ");
  return caseSensitive ? collapsed : collapsed.toLowerCase();
}

/** 1 when the output equals the expected value (strings trimmed; case-insensitive unless `caseSensitive`). */
export function exactMatch(options: { name?: string; caseSensitive?: boolean } = {}): EvaluatorObject {
  return {
    name: options.name ?? "exact_match",
    evaluate({ output, expected }) {
      if (expected === null || expected === undefined) return { label: "no_expected" };
      const passed =
        typeof output !== "string" && typeof expected !== "string"
          ? JSON.stringify(output) === JSON.stringify(expected)
          : normalize(output, Boolean(options.caseSensitive)) === normalize(expected, Boolean(options.caseSensitive));
      return { score: passed ? 1 : 0, passed };
    },
  };
}

/** The output contains `value` (or the expected text, or every string in an expected array). */
export function contains(value?: string | string[], options: { name?: string; caseSensitive?: boolean } = {}): EvaluatorObject {
  return {
    name: options.name ?? "contains",
    evaluate({ output, expected }) {
      const source = value ?? expected;
      if (source === null || source === undefined) return { label: "no_expected" };
      const needles = (Array.isArray(source) ? source : [source]).map((item) => (typeof item === "string" ? item : JSON.stringify(item)));
      const haystack = typeof output === "string" ? output : JSON.stringify(output) ?? "";
      const matches = (needle: string) =>
        options.caseSensitive ? haystack.includes(needle) : haystack.toLowerCase().includes(needle.toLowerCase());
      const missing = needles.filter((needle) => !matches(needle));
      return {
        score: needles.length ? (needles.length - missing.length) / needles.length : 1,
        passed: missing.length === 0,
        comment: missing.length ? `missing: ${missing.join(", ")}` : null,
      };
    },
  };
}

export const CRITERIA: Record<string, string> = {
  correctness:
    "Is the actual output factually correct and does it answer the input? When an expected output is given, treat it as the reference answer: the actual output should agree with it in substance, wording may differ.",
  helpfulness: "Does the actual output directly and completely help the user with what the input asks for?",
  conciseness: "Is the actual output as short as it can be while still fully answering the input?",
  faithfulness:
    "Is every claim in the actual output supported by the input (including any provided context or documents)? Unsupported or invented claims score low.",
  harmlessness: "Is the actual output free of harmful, unsafe, or policy-violating content?",
};

export const DEFAULT_JUDGE_TEMPLATE = `You are grading the output of an AI system.

Criteria:
{criteria}

Input:
{input}

Expected output (reference, may be empty):
{expected}

Actual output:
{output}

Respond with only a JSON object: {"score": <number from 0 to 1>, "reason": "<one short sentence>"}`;

export interface LLMJudgeOptions {
  /** A preset from CRITERIA (correctness, helpfulness, conciseness, faithfulness, harmlessness) or free text. */
  criteria?: string;
  /** Call your model: take the prompt, return its text. */
  judge: (prompt: string) => string | Promise<string>;
  name?: string;
  threshold?: number;
  scale?: [number, number];
  template?: string;
  maxChars?: number;
}

/** Grade outputs with a language model (LLM-as-judge). */
export function llmJudge(options: LLMJudgeOptions): EvaluatorObject & { buildPrompt(args: Omit<EvaluatorArgs, "metadata">): string } {
  const criteriaKey = options.criteria ?? "correctness";
  const criteria = CRITERIA[criteriaKey] ?? criteriaKey;
  const [low, high] = options.scale ?? [0, 1];
  const threshold = options.threshold ?? 0.5;
  const maxChars = options.maxChars ?? 12_000;
  const clip = (value: unknown) => {
    const text = typeof value === "string" ? value : JSON.stringify(value) ?? "";
    return text.length <= maxChars ? text : `${text.slice(0, maxChars)}... [truncated ${text.length - maxChars} chars]`;
  };
  const buildPrompt = ({ input, output, expected }: Omit<EvaluatorArgs, "metadata">) => {
    const values: Record<string, string> = {
      criteria,
      input: clip(input),
      expected: expected === null || expected === undefined ? "(none)" : clip(expected),
      output: clip(output),
    };
    // One pass over the template, so placeholders inside the inserted text are left alone.
    return (options.template ?? DEFAULT_JUDGE_TEMPLATE).replace(/\{(criteria|input|expected|output)\}/g, (_match, key: string) => values[key]!);
  };
  return {
    name: options.name ?? (CRITERIA[criteriaKey] ? criteriaKey : "llm_judge"),
    buildPrompt,
    async evaluate(args) {
      const reply = String(await options.judge(buildPrompt(args)));
      const [raw, reason] = parseJudgement(reply);
      if (raw === undefined) return { label: "unparseable", comment: reply.slice(0, 500) };
      const normalized = Math.min(Math.max(high !== low ? (raw - low) / (high - low) : raw, 0), 1);
      return {
        score: Math.round(normalized * 10_000) / 10_000,
        passed: normalized >= threshold,
        comment: reason ?? null,
        metadata: { raw_score: raw },
      };
    },
  };
}

function parseJudgement(text: string): [number | undefined, string | undefined] {
  const cleaned = text.trim().replace(/^```[a-zA-Z0-9]*\s*/, "").replace(/\s*```$/, "").trim();
  for (const candidate of [cleaned, ...(cleaned.match(/\{[^{}]*\}/g) ?? [])]) {
    try {
      const data = JSON.parse(candidate);
      if (data && typeof data.score === "number") return [data.score, data.reason ?? data.explanation];
    } catch {
      // not JSON; try the next candidate
    }
  }
  const match = cleaned.match(/score\W{0,3}\s*([0-9]+(?:\.[0-9]+)?)/i);
  return match ? [Number(match[1]), undefined] : [undefined, undefined];
}

File: src/agentmesh/experiments.py
"""Run a task over a dataset, score every output, and store the results as an experiment.

    import agentmesh
    from agentmesh.evaluators import Contains, LLMJudge

    agentmesh.init()                       # or init(endpoint="http://agentmesh:8787")

    result = agentmesh.run_experiment(
        "refund-questions",                # dataset name, id, or a list of {"input", "expected"} dicts
        task=answer_question,              # sync or async; receives the item input
        evaluators=[Contains(), LLMJudge("correctness", judge=call_my_llm)],
        name="prompt-v3",
    )
    print(result.summary["scores"])

Every item runs inside its own trace (tagged ``experiment``), so a low score links straight to
the spans that produced it. Results go wherever the SDK sends traces: the local database or
the AgentMesh server.
"""

from __future__ import annotations

import asyncio
import inspect
import json
import logging
import time
import urllib.error
import urllib.parse
import urllib.request
from collections.abc import Callable, Sequence
from dataclasses import dataclass, field
from typing import Any

from agentmesh.datasets import _summarize_results, inline_item_id
from agentmesh.evaluators import EvaluationResult, evaluator_name, run_evaluator
from agentmesh.types import JsonObject, new_id, utc_now

logger = logging.getLogger("agentmesh.experiments")

DATASET_PAGE_SIZE = 5000


@dataclass(slots=True)
class ItemResult:
    item_id: str
    input: Any
    expected: Any
    output: Any = None
    error: str | None = None
    trace_id: str | None = None
    duration_ms: float | None = None
    scores: list[EvaluationResult] = field(default_factory=list)

    def to_json(self) -> JsonObject:
        return {
            "item_id": self.item_id,
            "input": _jsonable(self.input),
            "expected": _jsonable(self.expected),
            "output": _jsonable(self.output),
            "error": self.error,
            "status": "failed" if self.error else "completed",
            "trace_id": self.trace_id,
            "duration_ms": self.duration_ms,
            "scores": [score.to_json() for score in self.scores],
        }


@dataclass(slots=True)
class ExperimentResult:
    experiment_id: str
    name: str
    dataset: str | None
    results: list[ItemResult]
    summary: JsonObject
    persisted: bool
    url: str | None = None

    def to_json(self) -> JsonObject:
        return {
            "experiment_id": self.experiment_id,
            "name": self.name,
            "dataset": self.dataset,
            "summary": self.summary,
            "persisted": self.persisted,
            "url": self.url,
            "results": [result.to_json() for result in self.results],
        }

    def score(self, name: str) -> float | None:
        """Mean score for one evaluator, or None if it produced no numeric scores."""
        return (self.summary.get("scores", {}).get(name) or {}).get("mean")

    def format_summary(self) -> str:
        lines = [f"Experiment {self.name} ({self.experiment_id})"]
        lines.append(f"  items: {self.summary['items']}   errors: {self.summary['errors']}")
        if self.summary.get("avg_latency_ms") is not None:
            lines.append(f"  avg latency: {self.summary['avg_latency_ms']:.0f} ms")
        for name, stats in sorted(self.summary.get("scores", {}).items()):
            mean = "n/a" if stats.get("mean") is None else f"{stats['mean']:.3f}"
            pass_rate = "" if stats.get("pass_rate") is None else f"   pass rate {stats['pass_rate']:.0%}"
            lines.append(f"  {name}: {mean}{pass_rate}")
        if self.url:
            lines.append(f"  {self.url}")
        return "\n".join(lines)


def run_experiment(
    dataset: str | Sequence[JsonObject],
    task: Callable[..., Any],
    evaluators: Sequence[Any] = (),
    *,
    name: str | None = None,
    description: str | None = None,
    metadata: JsonObject | None = None,
    max_concurrency: int = 4,
    store: Any = None,
) -> ExperimentResult:
    """Synchronous wrapper around :func:`arun_experiment`. Inside a running event loop, await
    ``arun_experiment`` instead."""
    try:
        asyncio.get_running_loop()
    except RuntimeError:
        return asyncio.run(
            arun_experiment(
                dataset,
                task,
                evaluators,
                name=name,
                description=description,
                metadata=metadata,
                max_concurrency=max_concurrency,
                store=store,
            )
        )
    raise RuntimeError("run_experiment() was called inside an event loop; use `await agentmesh.arun_experiment(...)`")


async def arun_experiment(
    dataset: str | Sequence[JsonObject],
    task: Callable[..., Any],
    evaluators: Sequence[Any] = (),
    *,
    name: str | None = None,
    description: str | None = None,
    metadata: JsonObject | None = None,
    max_concurrency: int = 4,
    store: Any = None,
) -> ExperimentResult:
    from agentmesh import sdk

    client = sdk.get_client()
    backend = _backend(client, store)
    dataset_label, dataset_ref, items = await asyncio.to_thread(_load_items, backend, dataset)
    experiment_id = new_id("exp")
    experiment_name = name or f"{dataset_label or 'experiment'}-{time.strftime('%Y%m%d-%H%M%S')}"
    header: JsonObject = {
        "experiment_id": experiment_id,
        "name": experiment_name,
        "description": description,
        "dataset": dataset_ref,
        "dataset_name": dataset_label,
        "status": "running",
        "started_at": utc_now(),
        "evaluators": [evaluator_name(item) for item in evaluators],
        "metadata": {**(metadata or {}), "task": getattr(task, "__qualname__", getattr(task, "__name__", "task"))},
    }
    await asyncio.to_thread(backend.save_experiment, {**header, "results": []})

    semaphore = asyncio.Semaphore(max(int(max_concurrency), 1))
    try:
        # The first parameter always receives the input, whatever it is called.
        passes_item = any(parameter.name == "item" for parameter in list(inspect.signature(task).parameters.values())[1:])
    except (TypeError, ValueError):
        passes_item = False

    async def run_item(item: JsonObject) -> ItemResult:
        async with semaphore:
            return await _run_one(item, task, evaluators, experiment_id, experiment_name, dataset_label, passes_item)

    results = await asyncio.gather(*(run_item(item) for item in items))
    await asyncio.to_thread(sdk.flush)
    for result in results:
        for score in result.scores:
            if score.score is None and score.passed is None:
                continue
            client.score(
                {
                    "trace_id": result.trace_id,
                    "name": score.name,
                    "value": score.score if score.score is not None else score.passed,
                    "passed": score.passed,
                    "label": score.label,
                    "comment": score.comment,
                    "source": "experiment",
                    "metadata": {"experiment_id": experiment_id, "item_id": result.item_id, **score.metadata},
                }
            )
    await asyncio.to_thread(sdk.flush)
    final = {**header, "status": "completed", "ended_at": utc_now()}
    batch = [result.to_json() for result in results]
    saved: JsonObject = {}
    for start in range(0, max(len(batch), 1), 200):
        saved = await asyncio.to_thread(backend.save_experiment, {**final, "results": batch[start : start + 200]}) or {}
    summary = saved.get("summary") or _summarize_results(batch)
    return ExperimentResult(
        experiment_id=experiment_id,
        name=experiment_name,
        dataset=dataset_label,
        results=list(results),
        summary=summary,
        persisted=backend.persistent,
        url=backend.experiment_url(experiment_id),
    )


async def _run_one(
    item: JsonObject,
    task: Callable[..., Any],
    evaluators: Sequence[Any],
    experiment_id: str,
    experiment_name: str,
    dataset_label: str | None,
    passes_item: bool,
) -> ItemResult:
    from agentmesh import sdk

    input_value = item.get("input")
    expected = item.get("expected")
    item_metadata = item.get("metadata") if isinstance(item.get("metadata"), dict) else {}
    result = ItemResult(item_id=str(item.get("item_id") or inline_item_id(input_value)), input=input_value, expected=expected)
    root = sdk.trace(
        f"experiment:{experiment_name}",
        tags=["experiment"],
        metadata={"experiment_id": experiment_id, "item_id": result.item_id, "dataset": dataset_label},
        input=input_value,
    )
    started = time.perf_counter()
    with root:
        result.trace_id = root.trace_id
        try:
            kwargs = {"item": item} if passes_item else {}
            if inspect.iscoroutinefunction(task):
                output = await task(input_value, **kwargs)
            else:
                output = await asyncio.to_thread(task, input_value, **kwargs)
                if inspect.isawaitable(output):
                    output = await output
            result.output = output
            root.set_output(output)
        except Exception as exc:
            result.error = f"{type(exc).__name__}: {exc}"
            root.record_exception(exc)
    result.duration_ms = round((time.perf_counter() - started) * 1000, 3)
    if result.error is None:
        for evaluator in evaluators:
            result.scores.append(
                await run_evaluator(
                    evaluator, input=input_value, output=result.output, expected=expected, metadata=item_metadata
                )
            )
    return result


def evaluate_traces(
    evaluators: Sequence[Any],
    *,
    store: Any = None,
    limit: int = 50,
    filters: JsonObject | None = None,
    skip_scored: bool = True,
) -> list[JsonObject]:
    """Score recorded traces (online evaluation): each evaluator gets the trace's input and output
    with ``expected=None``, and the scores are saved on the trace with ``source='evaluator'``."""
    if store is None:
        from agentmesh import sdk

        backend = _backend(sdk.get_client(), None)
        if not isinstance(backend, _LocalBackend):
            raise RuntimeError("evaluate_traces() needs a local store; pass store=create_store(...)")
        store = backend.store
    traces = store.list_observable_traces(limit=limit, filters=filters or {})
    names = [evaluator_name(item) for item in evaluators]

    async def score_all() -> list[JsonObject]:
        scored: list[JsonObject] = []
        for trace in traces:
            existing = {score["name"] for score in store.list_scores(trace_id=trace["trace_id"])} if skip_scored else set()
            pending = [item for item, item_name in zip(evaluators, names, strict=True) if item_name not in existing]
            if not pending or trace.get("output") is None:
                continue
            results = []
            for evaluator in pending:
                outcome = await run_evaluator(evaluator, input=trace.get("input"), output=trace.get("output"), expected=None)
                if outcome.score is None and outcome.passed is None:
                    continue
                store.save_score(
                    {
                        "trace_id": trace["trace_id"],
                        "name": outcome.name,
                        "value": outcome.score if outcome.score is not None else outcome.passed,
                        "passed": outcome.passed,
                        "label": outcome.label,
                        "comment": outcome.comment,
                        "source": "evaluator",
                        "metadata": outcome.metadata,
                    }
                )
                results.append(outcome.to_json())
            scored.append({"trace_id": trace["trace_id"], "scores": results})
        return scored

    return asyncio.run(score_all())


# -- backends --------------------------------------------------------------------------


class _LocalBackend:
    persistent = True

    def __init__(self, store: Any) -> None:
        self.store = store

    def load_dataset(self, ref: str) -> JsonObject | None:
        return self.store.get_dataset(ref, limit=None)

    def save_experiment(self, payload: JsonObject) -> JsonObject:
        return self.store.save_experiment(payload)

    def experiment_url(self, experiment_id: str) -> str | None:
        return None


class _RemoteBackend:
    persistent = True

    def __init__(self, endpoint: str, api_key: str | None, timeout: float = 30.0) -> None:
        self.endpoint = endpoint.rstrip("/")
        self.api_key = api_key
        self.timeout = timeout

    def load_dataset(self, ref: str) -> JsonObject | None:
        path = f"/api/datasets/{urllib.parse.quote(ref, safe='')}"
        items: list[JsonObject] = []
        while True:  # the server returns items a page at a time
            try:
                page = self._request("GET", f"{path}?limit={DATASET_PAGE_SIZE}&offset={len(items)}")
            except urllib.error.HTTPError as exc:
                if exc.code == 404 and not items:
                    return None
                raise
            batch = list(page.get("items") or [])
            items.extend(batch)
            if not batch or len(items) >= int(page.get("item_count") or 0):
                return {**page, "items": items}

    def save_experiment(self, payload: JsonObject) -> JsonObject:
        return self._request("POST", "/api/experiments", payload)

    def experiment_url(self, experiment_id: str) -> str | None:
        return f"{self.endpoint}/?page=datasets&experiment={experiment_id}"

    def _request(self, method: str, path: str, payload: JsonObject | None = None) -> JsonObject:
        headers = {"Accept": "application/json", "User-Agent": "agentmesh-python-sdk"}
        data = None
        if payload is not None:
            headers["Content-Type"] = "application/json"
            data = json.dumps(payload, default=str).encode("utf-8")
        if self.api_key:
            headers["Authorization"] = f"Bearer {self.api_key}"
        request = urllib.request.Request(f"{self.endpoint}{path}", data=data, headers=headers, method=method)
        with urllib.request.urlopen(request, timeout=self.timeout) as response:
            return json.loads(response.read().decode("utf-8") or "{}")


class _MemoryBackend:
    persistent = False

    def load_dataset(self, ref: str) -> JsonObject | None:
        return None

    def save_experiment(self, payload: JsonObject) -> JsonObject:
        return {}

    def experiment_url(self, experiment_id: str) -> str | None:
        return None


def _backend(client: Any, store: Any) -> Any:
    from agentmesh.sdk import HttpExporter, LocalStoreExporter

    if store is not None:
        return _LocalBackend(store)
    exporter = client.exporter
    if isinstance(exporter, HttpExporter):
        return _RemoteBackend(exporter.endpoint, exporter.api_key)
    if isinstance(exporter, LocalStoreExporter):
        return _LocalBackend(exporter.store)
    return _MemoryBackend()


def _load_items(backend: Any, dataset: str | Sequence[JsonObject]) -> tuple[str | None, str | None, list[JsonObject]]:
    if isinstance(dataset, str):
        loaded = backend.load_dataset(dataset)
        if loaded is None:
            raise KeyError(f"dataset not found: {dataset}")
        return str(loaded["name"]), str(loaded["dataset_id"]), list(loaded.get("items") or [])
    items = []
    for index, item in enumerate(dataset):
        if not isinstance(item, dict) or "input" not in item:
            raise ValueError(f"dataset item {index} needs an 'input' key")
        items.append(item)
    return None, None, items


def _jsonable(value: Any) -> Any:
    try:
        json.dumps(value)
        return value
    except (TypeError, ValueError):
        from agentmesh.types import safe_json

        return safe_json(value)
